### ISSUE TITLE
Load 3d TIFF images for CNN3D

### ISSUE THREAD

**Issue by AbdelmajidB** [eclipse/deeplearning4j#7948]
Hi, i'm working on 3d semantic segmentation based on CNN3D. I have o problem for load the third dimension for 3D features, for loading the 3d labels i created the following Label generator:
https://gist.github.com/AbdelmajidB/19a605acbc57eb8ad3173f9f2f63a29a
which load a 3D tiff using NativeImageLoader, the third dimension is loaded as a batch size and i reshape it to 5D NCDHW to support CNN3D and it works. but for features i try load 3d TIFF using  ImageRecordReader and RecordReader(Multi)DataSetIterator but it not loding the third dimension, for 32*32*32 3d image the showed features shape is : [1, 1, 32, 32]. Please any idea to resolve this isssue.

===
* Deeplearning4j version: 1.0.0-beta4
* Platform : Windows 7 
* CUDA version: 9.2
* NVIDIA driver version: 398.75

**Comment by saudet** [comment]
@AlexDBlack Should we consider this as a duplicate of #7689?

/cc @BioinfoTongLI

**Comment by AlexDBlack** [comment]
No, I think I'll keep this one open. That was mainly "multi-page tiff in NativeImageLoader".
This is more "CNN3D data from record reader". Obviously related, sure.

Anyway, I started adding extra constructor args into ImageRecordReader for this. I didn't like where it was going, so now I think we take a different approach to this...

We introduce a CNN3DRecordReader:
* Initially just support multi-page tiff using NativeImageLoader
* Design in such a way that later we can add video, [DICOM](https://en.wikipedia.org/wiki/DICOM), etc as supported formats. *Maybe* hdf5 also.
* Allow user to configure NCDHW vs. NDHWC

Though we could put the multi-page tiff support into ImageRecordReader, we'll then need extra args for depth, and then some of the other args (like image transform) won't apply. Seems like splitting it out to a separate reader would be much cleaner and more maintainable.

**Comment by AlexDBlack** [comment]
Also there's an issue at present with ImageRecordReader and loading multi-page tiffs that are a different size to the requested image size. My thinking there is we load the original size to INDArray, then use ResizeBilinear op in ND4J... unless there's a way to do that natively to the PIXA.

**Comment by AlexDBlack** [comment]
I've made some progress, but still far from ready:
https://github.com/SkymindIO/deeplearning4j/pull/56

I'm seeing a couple of weird things with tiff loading though... here's one page (one slice along depth dimension of cnn3d)
https://gist.github.com/AlexDBlack/86fc562a227aea3b02404e6d6c1a550d

There's 2 issues here: (see png)
1. Artifact in top left of some images - doesn't appear in the png (or when I view that page of the tiff in irfanview)
2. Different maximum values - max of 253 in PNG (seems reasonable), max of 177 from the tiff

https://github.com/deeplearning4j/dl4j-test-resources/tree/master/src/main/resources/datavec-data-image/multipage_tiff

All pages vs png - note tiff is first, png is second
https://gist.github.com/AlexDBlack/5e9bc4c324c49a8ef51cfffa4e78820b

As for how I made that tiff: irfanview -> multipage images -> create multipage tif, with default settings (no compression etc)

Any thoughts @saudet ?

**Comment by saudet** [comment]
Are you sure this isn't caused by "resize_bilinear"? You get the same thing with that commented out?

**Comment by AlexDBlack** [comment]
Yeah, that test case doesn't use resize bilinear, it's already the correct shape.
(I was seeing issues with that though, but figured we'd get the 'already correct shape' case working first)

**Comment by saudet** [comment]
I see what the issue is. The TIFF is 8 bps, but it's not grayscale, it's RGB with a palette:
```
$ file multipage_10_mnist_digits.tif
multipage_10_mnist_digits.tif: TIFF image data, little-endian, direntries=16, height=28, bps=8, compression=none, PhotometricIntepretation=RGB Palette, orientation=upper-left, width=28
```
So Leptonica loads that as RGB with 3 channels, by expanding each pixel to the color in the palette, but Leptonica basically adds a 0 alpha channel, and we get RGBA data actually, then ND4J converts the RGBA back to grayscale, but that's failing for some reason. Anyway, @BioinfoTongLI didn't make it to support multiple channels, so we should provide a grayscale TIFF file there instead for tests.
